src/core/Akka/Event/EventStream.cs

```diff
@@ -108,87 +114,69 @@ public override bool Unsubscribe(IActorRef subscriber)
+                            throw new IllegalActorStateException("EventStream is corrupted");
                     }
-                    v.Value.ForEach(RegisterWithUnsubscriber);
-
-
-                }
-                else
-                {
-                    InitUnsubscriber(unsubscriber, system);
                 }
+            }
+        }
 
-
-            }).With<Right<IActorRef>>(presentUnsubscriber =>
-            {
-                if (_debug)
-                {
-                    Publish(new Debug(SimpleName(this), GetType(),
-                        string.Format("not using unsubscriber {0}, because already initialized with {1}", unsubscriber, presentUnsubscriber)));
-
-                }
-            }).WasHandled;
+        [Obsolete("Should be removed in 1.5")]
+        public bool InitUnsubscriber(IActorRef unsubscriber, ActorSystem system)
+        {
+            StartUnsubscriber((ActorSystemImpl)system);
+            return true;
         }
 
         private void RegisterWithUnsubscriber(IActorRef subscriber)
         {
-            _initiallySubscribedOrUnsubscriber.Match().With<Left<IImmutableSet<IActorRef>>>(v =>
+            if (_unsubscriber.Value.IsNobody())
             {
-                if (_initiallySubscribedOrUnsubscriber.CompareAndSet(v,
-                    Either.Left<IImmutableSet<IActorRef>>(v.Value.Add(subscriber))))
-                {
-                    RegisterWithUnsubscriber(subscriber);
-                }
-
-            }).With<Right<IActorRef>>(unsubscriber =>
+                // pending
+                _pendingUnsubscribers.TryAdd(subscriber);
+            }
+            else
             {
-                unsubscriber.Value.Tell( new EventStreamUnsubscriber.Register(subscriber));
-            });
+                _unsubscriber.Value.Tell( new EventStreamUnsubscriber.Register(subscriber));
+            }
+            
         }
 
         private void UnregisterIfNoMoreSubscribedChannels(IActorRef subscriber)
         {
-            _initiallySubscribedOrUnsubscriber.Match().With<Left<IImmutableSet<IActorRef>>>(v =>
+            // not an important operation. If we fail to process this message due to a race condition, then the
+            // death watch subscription is a no-op anyway.
+            if (!_unsubscriber.Value.IsNobody())
             {
-                if (_initiallySubscribedOrUnsubscriber.CompareAndSet(v,
-                    Either.Left<IImmutableSet<IActorRef>>(v.Value.Remove(subscriber))))
-                {
-                    UnregisterIfNoMoreSubscribedChannels(subscriber);
-                }
-
-            }).With<Right<IActorRef>>(unsubscriber =>
-            {
-                unsubscriber.Value.Tell(new EventStreamUnsubscriber.UnregisterIfNoMoreSubscribedChannels(subscriber));
-            });
+                _unsubscriber.Value.Tell(new EventStreamUnsubscriber.UnregisterIfNoMoreSubscribedChannels(subscriber));
+            }
```

Comment:
No, because this event only gets fired when the actor is already unsubscribed to all topics. If the `_unsubscriber` hasn't been created yet then it can't have the deathwatch subscription that this message removes in the first place.